stress: failed test in cockroach/server/server.test: TestStatusSummaries

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/9ef38d72441225bc3c665af3ae4a2b74d64d93d5

Stress build found a failed test:

```
=== RUN   TestStatusSummaries
I161014 05:01:43.778614 4777 storage/engine/rocksdb.go:349  opening in memory rocksdb instance
I161014 05:01:43.781934 4777 gossip/gossip.go:234  [n?] initial resolvers: []
W161014 05:01:43.781956 4777 gossip/gossip.go:1045  [n?] no resolvers found; use --join to specify a connected node
W161014 05:01:43.783401 4777 server/status/runtime.go:116  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I161014 05:01:43.784213 4777 server/node.go:415  [n?] store [n0,s0] not bootstrapped
I161014 05:01:43.786144 4983 storage/replica_trigger.go:258  [s1,r1:/M{in-ax},r1:/M{in-ax}] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 410117h1m52.785514746s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=2016-10-14 05:01:43.786096493 +0000 UTC]
I161014 05:01:43.786862 4777 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I161014 05:01:43.786970 4777 server/node.go:347  [n?] **** cluster {b70c88b0-cb03-4c57-a53a-5da1265f23d2} has been created
I161014 05:01:43.786991 4777 server/node.go:348  [n?] **** add additional nodes by specifying --join=127.0.0.1:42149
I161014 05:01:43.787331 4777 gossip/gossip.go:268  [n?] NodeID set to 1
I161014 05:01:43.787974 4777 storage/store.go:1075  [n?] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I161014 05:01:43.788002 4777 server/node.go:428  [n?] initialized store [n1,s1]: {Capacity:536870912 Available:536870912 RangeCount:0}
I161014 05:01:43.788029 4777 server/node.go:316  [n?] node ID 1 initialized
I161014 05:01:43.788084 4777 gossip/gossip.go:282  [n?] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:42149" > attrs:<> locality:<> 
I161014 05:01:43.788151 4777 storage/stores.go:292  [n?] read 0 node addresses from persistent storage
I161014 05:01:43.788194 4777 server/node.go:558  [n?] connecting to gossip network to verify cluster ID...
I161014 05:01:43.788218 4777 server/node.go:579  [n?] node connected via gossip and verified as part of cluster {"b70c88b0-cb03-4c57-a53a-5da1265f23d2"}
I161014 05:01:43.788244 4777 server/node.go:366  [n?] node=1: started with [[]=] engine(s) and attributes []
I161014 05:01:43.788271 4777 server/server.go:543  [n1] starting https server at 127.0.0.1:43712
I161014 05:01:43.788282 4777 server/server.go:544  [n1] starting grpc/postgres server at 127.0.0.1:42149
I161014 05:01:43.788292 4777 server/server.go:545  [n1] advertising CockroachDB node at 127.0.0.1:42149
I161014 05:01:43.788980 4950 storage/split_queue.go:103  [split,n1,s1,r1:/M{in-ax}] splitting at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
I161014 05:01:43.807283 4950 storage/replica_command.go:2336  [split,n1,s1,r1:/M{in-ax},r1:/M{in-ax}] initiating a split of this range at key /Table/11 [r2]
I161014 05:01:43.810151 4950 storage/replica_command.go:2336  [split,n1,s1,r1:/{Min-Table/11},r2:/{Table/11-Max}] initiating a split of this range at key /Table/12 [r3]
I161014 05:01:43.812945 4950 storage/replica_command.go:2336  [split,n1,s1,r1:/{Min-Table/11},r3:/{Table/12-Max}] initiating a split of this range at key /Table/13 [r4]
I161014 05:01:43.816303 4950 storage/replica_command.go:2336  [split,n1,s1,r1:/{Min-Table/11},r4:/{Table/13-Max}] initiating a split of this range at key /Table/14 [r5]
I161014 05:01:43.848560 4777 storage/replica_command.go:2336  [n1,s1,r1:/{Min-Table/11}] initiating a split of this range at key "b" [r6]
I161014 05:01:43.854868 4777 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I161014 05:01:43.855016 5069 http2_client.go:1053  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I161014 05:01:43.855025 3338 http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:42149->127.0.0.1:44226: use of closed network connection
I161014 05:01:43.855154 5063 /go/src/google.golang.org/grpc/clientconn.go:667  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:42149: operation was canceled"; Reconnecting to {"127.0.0.1:42149" <nil>}
I161014 05:01:43.855171 5063 /go/src/google.golang.org/grpc/clientconn.go:767  grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing
--- FAIL: TestStatusSummaries (0.08s)
    node_test.go:432: 3: livebytes is not greater than or equal to expected value.
        Expected 23585.000000, Actual 23543.000000
```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
16 runs so far, 0 failures, over 15s
16 runs so far, 0 failures, over 20s
32 runs so far, 0 failures, over 25s
32 runs so far, 0 failures, over 30s
48 runs so far, 0 failures, over 35s
48 runs so far, 0 failures, over 40s
64 runs so far, 0 failures, over 45s
64 runs so far, 0 failures, over 50s
80 runs so far, 0 failures, over 55s
80 runs so far, 0 failures, over 1m0s
96 runs so far, 0 failures, over 1m5s
96 runs so far, 0 failures, over 1m10s
111 runs so far, 0 failures, over 1m15s
112 runs so far, 0 failures, over 1m20s
125 runs so far, 0 failures, over 1m25s
128 runs so far, 0 failures, over 1m30s
138 runs so far, 0 failures, over 1m35s
144 runs so far, 0 failures, over 1m40s
149 runs so far, 0 failures, over 1m45s
160 runs so far, 0 failures, over 1m50s
163 runs so far, 0 failures, over 1m55s
176 runs so far, 0 failures, over 2m0s
177 runs so far, 0 failures, over 2m5s
192 runs so far, 0 failures, over 2m10s
193 runs so far, 0 failures, over 2m15s
208 runs so far, 0 failures, over 2m20s
209 runs so far, 0 failures, over 2m25s
223 runs so far, 0 failures, over 2m30s
225 runs so far, 0 failures, over 2m35s
239 runs so far, 0 failures, over 2m40s
241 runs so far, 0 failures, over 2m45s
254 runs so far, 0 failures, over 2m50s
256 runs so far, 0 failures, over 2m55s
268 runs so far, 0 failures, over 3m0s
272 runs so far, 0 failures, over 3m5s
281 runs so far, 0 failures, over 3m10s
288 runs so far, 0 failures, over 3m15s
296 runs so far, 0 failures, over 3m20s
304 runs so far, 0 failures, over 3m25s
311 runs so far, 0 failures, over 3m30s
320 runs so far, 0 failures, over 3m35s
327 runs so far, 0 failures, over 3m40s
336 runs so far, 0 failures, over 3m45s
341 runs so far, 0 failures, over 3m50s
352 runs so far, 0 failures, over 3m55s
355 runs so far, 0 failures, over 4m0s
368 runs so far, 0 failures, over 4m5s
370 runs so far, 0 failures, over 4m10s
384 runs so far, 0 failures, over 4m15s
386 runs so far, 0 failures, over 4m20s
400 runs so far, 0 failures, over 4m25s
402 runs so far, 0 failures, over 4m30s
415 runs so far, 0 failures, over 4m35s
417 runs so far, 0 failures, over 4m40s
428 runs so far, 0 failures, over 4m45s
433 runs so far, 0 failures, over 4m50s
443 runs so far, 0 failures, over 4m55s
448 runs so far, 0 failures, over 5m0s
459 runs so far, 0 failures, over 5m5s
464 runs so far, 0 failures, over 5m10s
474 runs so far, 0 failures, over 5m15s
480 runs so far, 0 failures, over 5m20s
488 runs so far, 0 failures, over 5m25s
496 runs so far, 0 failures, over 5m30s
504 runs so far, 0 failures, over 5m35s
512 runs so far, 0 failures, over 5m40s
519 runs so far, 0 failures, over 5m45s
528 runs so far, 0 failures, over 5m50s
532 runs so far, 0 failures, over 5m55s
544 runs so far, 0 failures, over 6m0s
547 runs so far, 0 failures, over 6m5s
560 runs so far, 0 failures, over 6m10s
563 runs so far, 0 failures, over 6m15s
575 runs so far, 0 failures, over 6m20s
578 runs so far, 0 failures, over 6m25s
589 runs so far, 0 failures, over 6m30s
593 runs so far, 0 failures, over 6m35s
604 runs so far, 0 failures, over 6m40s
609 runs so far, 0 failures, over 6m45s
619 runs so far, 0 failures, over 6m50s
624 runs so far, 0 failures, over 6m55s
632 runs so far, 0 failures, over 7m0s
640 runs so far, 0 failures, over 7m5s
646 runs so far, 0 failures, over 7m10s
656 runs so far, 0 failures, over 7m15s
661 runs so far, 0 failures, over 7m20s
672 runs so far, 0 failures, over 7m25s
676 runs so far, 0 failures, over 7m30s
688 runs so far, 0 failures, over 7m35s
691 runs so far, 0 failures, over 7m40s
703 runs so far, 0 failures, over 7m45s
705 runs so far, 0 failures, over 7m50s
717 runs so far, 0 failures, over 7m55s
721 runs so far, 0 failures, over 8m0s
732 runs so far, 0 failures, over 8m5s
736 runs so far, 0 failures, over 8m10s
747 runs so far, 0 failures, over 8m15s
752 runs so far, 0 failures, over 8m20s
760 runs so far, 0 failures, over 8m25s
768 runs so far, 0 failures, over 8m30s
775 runs so far, 0 failures, over 8m35s
784 runs so far, 0 failures, over 8m40s
788 runs so far, 0 failures, over 8m45s
800 runs so far, 0 failures, over 8m50s
804 runs so far, 0 failures, over 8m55s
815 runs so far, 0 failures, over 9m0s
818 runs so far, 0 failures, over 9m5s
830 runs so far, 0 failures, over 9m10s
832 runs so far, 0 failures, over 9m15s
844 runs so far, 0 failures, over 9m20s
848 runs so far, 0 failures, over 9m25s
860 runs so far, 0 failures, over 9m30s
864 runs so far, 0 failures, over 9m35s
873 runs so far, 0 failures, over 9m40s
880 runs so far, 0 failures, over 9m45s
888 runs so far, 0 failures, over 9m50s
896 runs so far, 0 failures, over 9m55s
902 runs so far, 0 failures, over 10m0s
912 runs so far, 0 failures, over 10m5s
916 runs so far, 0 failures, over 10m10s
927 runs so far, 0 failures, over 10m15s
930 runs so far, 0 failures, over 10m20s
942 runs so far, 0 failures, over 10m25s
945 runs so far, 0 failures, over 10m30s
957 runs so far, 0 failures, over 10m35s
960 runs so far, 0 failures, over 10m40s
972 runs so far, 0 failures, over 10m45s
976 runs so far, 0 failures, over 10m50s
985 runs so far, 0 failures, over 10m55s
992 runs so far, 0 failures, over 11m0s
1001 runs so far, 0 failures, over 11m5s
1008 runs so far, 0 failures, over 11m10s
1016 runs so far, 0 failures, over 11m15s
1024 runs so far, 0 failures, over 11m20s
1028 runs so far, 0 failures, over 11m25s
1039 runs so far, 0 failures, over 11m30s
1041 runs so far, 0 failures, over 11m35s
1054 runs so far, 0 failures, over 11m40s
1056 runs so far, 0 failures, over 11m45s
1068 runs so far, 0 failures, over 11m50s
1072 runs so far, 0 failures, over 11m55s
1083 runs so far, 0 failures, over 12m0s
1088 runs so far, 0 failures, over 12m5s
1098 runs so far, 0 failures, over 12m10s
1104 runs so far, 0 failures, over 12m15s
1111 runs so far, 0 failures, over 12m20s
1120 runs so far, 0 failures, over 12m25s
1124 runs completed, 1 failures, over 12m29s
FAIL
```

Please assign, take a look and update the issue accordingly.

This does appear to be an asynchronous background goroutine writing a transaction, this time it's the liveness table.

For posterity, let me record the logging I added to find this (which is not appropriate to check in, even at a high verbosity):

TestStatusSummaries verifies that the values recorded in a metrics summary change predictably in relation to certain database activities, such as a write.  This is done by generating a "base" metrics summary, then performing the action, then generating another metrics summary and asserting that the difference between the two summaries makes sense.  However, an issue occurs when commands *other* than the one being tested occur on a background goroutine - previously this occurred with the EventLog, now it is happening with the Liveness table.  The unexpected command can alter the metrics summaries being evaluated, typically by making the LiveBytes count lower in the second summary - it is purely a race condition.

I discover the culprit background routine by adding the following logging:

+ In "TestStatusSummaries" itself, add log messages before and after status summaries are generated. This is currently added to the `forceWriteSummaries` method.  By doing this, we can more clearly gauge in the test output exactly when summaries are generated.
+ Add a log message to replicas wherever store-level MVCC stats are changed.  Currently this is done in three places - when snapshots are applied, when replicas are added, and most importantly after any proposed command is accepted and evaluated.  For example, the log message for an evaluated command looks like this:

```go
	if rpd.Delta.ValBytes != 0 {
		log.Warningf(
			ctx,
			"!!! command %T:: %+v, valbytes diff %+d",
			cmd.Requests[0].GetValue(),
			cmd.Requests[0].GetValue(),
			rpd.Delta.ValBytes,
		)
	}
    // Update store-level MVCC stats with merged range stats.
	r.store.metrics.addMVCCStats(rpd.Delta)
```

This log message files whenever ValBytes (which is the metric currently causing TestStatusSummaries to fail) is changed by a command. It outputs the type of the command, a struct representation of the command (which includes the keys affected), and the total change in bytes.  Note that because it reflects on one specific field of the delta, this is too specific to include permanently in the code, even at a high verbosity level.

Put together, the current log output I get from a failed `make stress` run looks like this:

```
W161110 14:38:00.294627 95 storage/replica_proposal.go:438  [n1,s1,r1/1:/{Min-Table/11}] !!! command *roachpb.ResolveIntentRequest, key range header:<key:"\003\225" > intent_txn:<id:<82e94cbc-12dc-45b0-a724-00be25c011fd> isolation:SERIALIZABLE key:"\001k\022\224\000\001rdsc" epoch:0 timestamp:<wall_time:1478806680290933640 logical:0 > priority:873817 sequence:2 batch_index:0 > status:COMMITTED poison:false , valbytes diff -156
I161110 14:38:00.294737 116 storage/replica_command.go:2360  [n1,split,s1,r1/1:/{Min-Table/11}] initiating a split of this range at key /Table/14 [r5]
W161110 14:38:00.296211 89 storage/replica_proposal.go:438  [n1,split,s1,r1/1:/{Min-Table/11}] !!! command *roachpb.NoopRequest, key range , valbytes diff +203
I161110 14:38:00.297162 97 storage/store.go:1713  [n1,split,s1,r1/1:/{Min-Table/11}] !!! updating MVCC Stats from split, valbytes diff 0
W161110 14:38:00.298093 99 storage/replica_proposal.go:438  [n1,s1,r1/1:/{Min-Table/11}] !!! command *roachpb.ResolveIntentRequest, key range header:<key:"\003\226" > intent_txn:<id:<d18bc9ba-04d5-44c2-af26-6b652565cb7e> isolation:SERIALIZABLE key:"\001k\022\225\000\001rdsc" epoch:0 timestamp:<wall_time:1478806680294784897 logical:0 > priority:152139 sequence:2 batch_index:0 > status:COMMITTED poison:false , valbytes diff -156
begin write statuses
W161110 14:38:00.330347 102 storage/replica_proposal.go:438  [n1,hb,s1,r1/1:/{Min-Table/11}] !!! command *roachpb.BeginTransactionRequest, key range header:<key:"\004\000liveness-\211" > , valbytes diff +103
W161110 14:38:00.332678 91 storage/replica_proposal.go:438  [n1,hb,s1,r1/1:/{Min-Table/11}] !!! command *roachpb.ConditionalPutRequest, key range header:<key:"\004\000liveness-\211" > value:<raw_bytes:"\330N\361,\003\010\001\020\001\032\014\010\206\237\240\355\224\333\361\302\024\020\000" timestamp:<wall_time:0 logical:0 > > blind:false , valbytes diff -80
end write statuses
grabbing node statuses
W161110 14:38:00.334519 103 storage/replica_proposal.go:438  [n1,s1,r1/1:/{Min-Table/11}] !!! command *roachpb.PutRequest, key range header:<key:"a" > value:<raw_bytes:"\022uD\251\003junk" timestamp:<wall_time:0 logical:0 > > inline:false blind:false , valbytes diff +9
W161110 14:38:00.334940 98 storage/replica_proposal.go:438  [n1,s1,r1/1:/{Min-Table/11}] !!! command *roachpb.PutRequest, key range header:<key:"c" > value:<raw_bytes:"_\275\345\242\003junk" timestamp:<wall_time:0 logical:0 > > inline:false blind:false , valbytes diff +9
begin write statuses
end write statuses
grabbing node statuses
I161110 14:38:00.339076 34 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161110 14:38:00.339267 147 http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:64406->127.0.0.1:64409: use of closed network connection
--- FAIL: TestStatusSummaries (0.08s)
	node_test.go:443: 2: livebytes is not greater than or equal to expected value.
		Expected 33050.000000, Actual 33015.000000
	node_test.go:443: 2: valbytes is not greater than or equal to expected value.
		Expected 32517.000000, Actual 32454.000000
```

Notice the BeginTransactionRequest and ConditionalPut on the liveness table, which overlaps with the first status summary being generated.  This throws off the expectations of the test, and causes it to fail.

The solution previously (when the overlapping transaction was around Node statuses) was to simply disable that subsystem; I don't know if that's possible with the liveness table, but I will investigate.

--- a/pkg/server/node_test.go
+++ b/pkg/server/node_test.go
@@ -457,9 +457,6 @@ func compareNodeStatus(
 				"replicas.leaseholders",
 			},
 			[]string{
-				"livebytes",
-				"keybytes",
-				"valbytes",
 				"livecount",
 				"keycount",
 				"valcount",

